runtime_scan/pkg/provider/docker/client.go

```diff
@@ -0,0 +1,585 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package docker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	types "github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/docker/api/types/volume"
+	"github.com/docker/docker/client"
+	"time"
+
+	"github.com/openclarity/vmclarity/api/models"
+	"github.com/openclarity/vmclarity/runtime_scan/pkg/provider"
+	"github.com/openclarity/vmclarity/shared/pkg/utils"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+type Client struct {
+	dockerClient *client.Client
+	config       *Config
+}
+
+func New(ctx context.Context) (*Client, error) {
+	config, err := NewConfig()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration. Provider=%s: %w", models.Docker, err)
+	}
+
+	if err = config.Validate(); err != nil {
+		return nil, fmt.Errorf("failed to validate provider configuration. Provider=%s: %w", models.Docker, err)
+	}
+
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return nil, fmt.Errorf("failed to load provider configuration. Provider=%s: %w", models.Docker, err)
+	}
+
+	return &Client{
+		dockerClient: dockerClient,
+		config:       config,
+	}, nil
+}
+
+func (c Client) Kind() models.CloudProvider {
+	return models.Docker
+}
+
+func (c Client) DiscoverAssets(ctx context.Context) ([]models.AssetType, error) {
+	var err error
+	wg := &sync.WaitGroup{}
+	errs := make(chan error, 2)
+	assets := make(chan models.AssetType)
+
+	wg.Add(1)
+	go c.getImages(ctx, assets, errs, wg)
+
+	wg.Add(1)
+	go c.getContainers(ctx, assets, errs, wg)
+
+	go func() {
+		wg.Wait()
+		close(errs)
+		close(assets)
+	}()
+
+	var ret []models.AssetType
+	for t := range assets {
+		ret = append(ret, t)
+	}
+
+	for e := range errs {
+		if e != nil {
+			// nolint:typecheck
+			err = errors.Join(err, e)
+		}
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return ret, nil
+}
+
+func (c Client) getImages(ctx context.Context, assets chan models.AssetType, errs chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	images, err := c.dockerClient.ImageList(ctx, types.ImageListOptions{})
+	if err != nil {
+		errs <- provider.FatalError{
+			Err: fmt.Errorf("failed to get images. Provider=%s: %w", models.Docker, err),
+		}
+		return
+	}
+
+	for _, i := range images {
+		asset, err := c.getAssetFromImage(ctx, i.ID)
+		if err != nil {
+			errs <- provider.FatalError{
+				Err: fmt.Errorf("failed to create AssetType from ContainerImageInfo. Provider=%s: %w", models.Docker, err),
+			}
+			return
+		}
+		assets <- asset
+	}
+}
+
+func (c Client) getContainers(ctx context.Context, assets chan models.AssetType, errs chan error, wg *sync.WaitGroup) {
+	defer wg.Done()
+
```

Comment:
Same notes like to getImages apply:
1. create separate single-fetch method
2. use that instead of doing everything in `getContainers`
3. get rid of provider-errors
4. remove unnecessary wg control